App Engine Service Split Traffic resource

khanali21 opened the pull request:
* Added custom_create in custom_code.
* Added appengine service resource

<!-- 
Note: You may see "This branch is out-of-date with the base branch"
when you submit a pull request. This is fine! We don't use the GitHub
merge button to merge PRs, and you can safely ignore that message.

Thanks for contributing!
-->

<!-- CHANGELOG for Downstream PRs.
EXTERNAL CONTRIBUTORS: Your reviewer will most likely fill this in for you, so don't worry about this section!

For some repos (currently Terraform GA/beta providers), we have the
ability to autogenerate CHANGELOGs.

Fill in the following release note code block to have it be added to the CHANGELOG, or leave the block empty if you don't expect this to be added to a downstream PR (i.e. docs-only changes or non-user facing changes)

Please also add any of the following appropriate labels to the PR:
- changelog: bugfix
- changelog: new-resource
- changelog: new-datasource
- changelog: deprecation
- changelog: breaking-change
-->
# Release Note for Downstream PRs (will be copied)

```release-note:new-resource
`google_app_engine_service_split_traffic`
```

modular-magician commented:
Hello!  I am a robot who works on Magic Modules PRs.

I have detected that you are a community contributor, so your PR will be assigned to someone with a commit-bit on this repo for initial review.  They will authorize it to run through our CI pipeline, which will generate downstream PRs.

Thanks for your contribution!  A human will be with you soon.

@emilymye, please review this PR or find an appropriate assignee.

khanali21 commented:
@rileykarson I noticed that there is no option for the custom_create code in the generator. I thought it fitting to add this. This helps in Appengine Service Resource which does not have Create method.

If custom_create was not deliberately added, then I can move the resource under third_party as the custom resource.

Thanks
Ali

khanali21 commented:
Example template to follow.

rileykarson commented:
For Terraform, I think we want to expose this under another name than `Service`. Since a service can't be (directly) created, I think it would be confusing to have a `google_app_engine_service` resource defined that depended on a version creating the service.


As a user, I'd expect my config to look like:

```hcl
resource "google_app_engine_service" "frontend" {
  id = "my-frontend-service"
  
} 

resource "google_app_engine_standard_app_version" "frontend_v2" {
  version_id = "v2"
  service = google_app_engine_service.frontend.id

  ...
}
```

When in reality, it's more likely to be like this, because the dependency is flipped:

```hcl
resource "google_app_engine_standard_app_version" "frontend_v2" {
  version_id = "v2"
  service = "my-frontend-service"

  ...
}

resource "google_app_engine_service" "frontend" {
  id = google_app_engine_standard_version.frontend_v2.service
  
} 
```

Instead, I'd lean towards exposing the resource as `google_app_engine_service_traffic_split`. I *believe* we can override the `name` value in `terraform.yaml` to accomplish that, but if not we'll need to add a duplicate entry in `api.yaml`.

khanali21 commented:
@rileykarson  Struggling with few things:
1. The create_verb can only be PUT and POST. Do you think I should modify ruby code to allow PATCH as the create_verb too.
2. The update_mask true includes the id field as well, however I want only one field "split" to be part of the updateMask, i.e. updateMask=split
3. I don't see the need for overriding name, as I changed the name of the resource and then also updated the resourceref for the Version resource. Let me know if there is any other need for name override. BTW I could not find how override the name, I did find legacy_name but could not make it work as intended. Can you tell me if that is the right one.


thanks

khanali21 commented on products/appengine/api.yaml:
```diff
@@ -73,24 +73,58 @@ objects:
+        required: false
+        properties:      
+          - !ruby/object:Api::Type::Enum
+            name: 'shardBy'
```
I am going to test it shortly

rileykarson replied:
Did you get a chance to test?

khanali21 commented on products/appengine/terraform.yaml:
```diff
@@ -64,7 +64,16 @@ overrides: !ruby/object:Overrides::ResourceOverrides
+    import_format: ["apps/{{project}}/services/{{service_id}}"]
+    mutex: "apps/{{project}}/services/{{service_id}}"
+    custom_code: !ruby/object:Provider::Terraform::CustomCode
+      custom_create: templates/terraform/custom_create/noop_on_create_appengine_version.go.erb
```
Not sure whether I need to add create_encoder in this case. So there are two things:
1. Having updateMask in the create_url, I think this can be handled by just adding the constant like:
```
    create_url: 'apps/{{project}}/services/{{service_id}}?{{migrateTraffic}}&updateMask=split'
````
2. Identifying if we need to add shardBy also in the updateMask (I am looking into it)

khanali21 commented:
@rileykarson  Kindly review this. I have made all the required changes, One question is still open.
Q. Should I just override the delete with no op as there is no deletion applicable in this resource.

khanali21 commented:
```
resource "google_storage_bucket" "bucket" {
	name = "<%= ctx[:vars]['bucket_name'] %>"
}

resource "google_storage_bucket_object" "object" {
	name   = "hello-world.zip"
	bucket = "${google_storage_bucket.bucket.name}"
	source = "./test-fixtures/appengine/hello-world.zip"
}

resource "google_app_engine_standard_app_version" "myapp_v1" {
  version_id = "v1"
  service = "myapp2"
  runtime = "nodejs10"
  noop_on_destroy = true
  entrypoint {
    shell = "node ./app.js"
  }
  deployment {
    zip {
      source_url = "https://storage.googleapis.com/${google_storage_bucket.bucket.name}/hello-world.zip"
    }  
  }
  env_variables = {
    port = "8080"
  } 

}
resource "google_app_engine_standard_app_version" "myapp_v2" {
  version_id = "v2"
  service = "myapp2"
  runtime = "nodejs10"
  noop_on_destroy = true
  entrypoint {
    shell = "node ./app.js"
  }
  deployment {
    zip {
      source_url = "https://storage.googleapis.com/${google_storage_bucket.bucket.name}/hello-world.zip"
    }  
  }
  env_variables = {
    port = "8080"
  } 

}

resource "google_app_engine_service_split_traffic" "myapp2" {
  service_id = "${google_app_engine_standard_app_version.myapp_v2.service}"
  migrate_traffic = false
  split {
    shard_by = "IP"
    allocations = {
      v1 = 0.75
      v2 = 0.25
    }
  }
}
```

rileykarson commented on products/appengine/api.yaml:
```diff
@@ -67,30 +67,74 @@ objects:
           - DENY
   - !ruby/object:Api::Resource
-    name: 'Service'
+    name: 'ServiceSplitTraffic'
```
Instead of renaming `Service`, would you mind duplicating this into a separate resource definition? It feels a little silly, but it's the best way to represent fine-grained resources right now.

rileykarson replied:
This will also require exclusions in Ansible / InSpec for the Magician to work.

rileykarson commented on products/appengine/api.yaml:
```diff
@@ -67,30 +67,74 @@ objects:
           Full path to the Service resource in the API. Example apps/myapp/services/default.
           This field is used in responses only. Any value specified here in a request is ignored.
       - !ruby/object:Api::Type::String
         name: 'id'
```
We have some liberty in terms of what we can call this field because it doesn't actually correspond to any fields we care about in the API. What do you think of `service` instead of `id`/`service_id`?

rileykarson replied:
wdyt here?

khanali21 commented:
@rileykarson Please review now.

Also here is my loud thinking on this:

Use case 1:
a) Enable appengine using google_app_engine_application resource
b) Create a new service named e.g. "myapp" and upload the code as v1 using google_app_engine_standard_app_version resource. (Please note that the service is automatically created if it is not existing). Use the noop_on_destroy to ensure that this version does not get deleted.
c) Create another version v2 for service myapp using google_app_engine_standard_app_version resource.
d) Split the traffic using the google_app_engine_service_split_traffic resource.

Problem: In this use case terraform destroy will leave the service myapp and version v1 when destroying all the resources.

Solution: Use the google_app_engine_service resource to delete the final version. For this we shall need the google_app_engine_service resource as I did in the first iteration which will ignore the create but allow deletion.

e) Destr

khanali21 commented:
@rileykarson Can you please review this. And give your comments.

rileykarson commented on templates/terraform/examples/app_engine_service_split_traffic.tf.erb:
```diff
@@ -0,0 +1,69 @@
+  split {
+    shard_by = "IP"
+    allocations = {
+      v1 = 0.75
```
Are you able to interpolate the version names off of the `google_app_engine_standard_app_version` resource?

rileykarson replied:
This is unaddressed.

rileykarson replied:
Resolving as old.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
This PR seems not to have generated downstream PRs before, as of b044582b7c95646194870fa3b6a4d9ea0905bd3d. 
## Pull request statuses
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I built this PR into one or more new PRs on other repositories, and when those are closed, this PR will also be merged and closed.
depends: https://github.com/terraform-providers/terraform-provider-google-beta/pull/1135 
depends: https://github.com/GoogleCloudPlatform/terraform-google-conversion/pull/199 
depends: https://github.com/terraform-providers/terraform-provider-google/pull/4451

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, b044582b7c95646194870fa3b6a4d9ea0905bd3d. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

rileykarson commented on provider/terraform/custom_code.rb:
```diff
@@ -96,6 +96,10 @@ class CustomCode < Api::Object
       # method's function header can't be changed, the template
       # inserts that for you - do not include it in your custom code.
       attr_reader :custom_delete
+      # This code replaces the entire create method.  Since the create
```
As long as we're not using this, do you mind removing it + the changes in `resource.erb`? I'd rather not add unused code.

rileykarson commented:
Just a heads up & to be explicit, waiting on https://github.com/GoogleCloudPlatform/magic-modules/pull/2331 to get merged before making another pass here.

khanali21 commented:
@rileykarson that's right. 
A quick question. I am getting this error when running the test
				Resource specified by ResourceName couldn't be found: google_app_engine_service_split_traffic."

I believe this may be expected, so I thought I should put
				ExpectError:             regexp.MustCompile("Resource specified by ResourceName couldn't be found: google_app_engine_service_split_traffic."),

is that right approach?
Any pointer for a custom resource like split traffic.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, e0ba30ea0617a09516630fa0b308669b9777b209. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

rileykarson commented on third_party/terraform/tests/resource_app_engine_service_split_traffic_test.go:
```diff
@@ -0,0 +1,128 @@
+				ImportState:             true,
+				ImportStateVerify:       true,
+				ImportStateVerifyIgnore: []string{"service_id", "split"},
+				ExpectError:             regexp.MustCompile("Resource specified by ResourceName couldn't be found: google_app_engine_service_split_traffic."),
```
I don't believe this will get evaluated during an `ImportState` step, only a `Config` one.

khanali21 replied:
yes, it does fail if I remove the line 35, which ExpectError. I did try to get to the bottom of this, but honestly could not understand. During debugging I found that it is failing after the test is finished, let me send you completed logs.

rileykarson replied:
Do you mind pushing this with the line removed? I'm curious why it's failing, and I'd like the test to be working before we merge this. No need for logs in that case, I can generate them myself.

khanali21 replied:
Thanks for your review @rileykarson , please check this.

rileykarson replied:
Can you remove this? I'd like to get this working (and ideally, generate the test)

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 7c017010a2df7b809768693e7300f28d4051a88a. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

rileykarson commented on third_party/terraform/tests/resource_app_engine_service_split_traffic_test.go:
```diff
@@ -0,0 +1,127 @@
+	"strings"
+	"testing"
+
+	"github.com/hashicorp/terraform/helper/acctest"
```
```suggestion
	"github.com/hashicorp/terraform-plugin-sdk/helper/acctest"
```

rileykarson replied:
This + the other helper import need to be fixed to use the new plugin SDK package.

rileykarson commented on third_party/terraform/tests/resource_app_engine_service_split_traffic_test.go:
```diff
@@ -0,0 +1,127 @@
+
+	"github.com/hashicorp/terraform/helper/acctest"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/terraform"
```
Is this import needed?

khanali21 commented:
@rileykarson I am unable to compile the code. I will try again today with clean setup.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 1779cbd547b751ee09ad89d1b949df63785617a8. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

khanali21 commented:
@rileykarson I am back working on this PR.  Can you do a quick review and let me know what it will take to get this merged. thanks in advance. 
I have rebased it on the latest master.

khanali21 commented:
@rileykarson So got over the issue with the compilation and also rebased it with the latest master. 
Will appreciate if we can get this merged as soon as possible.

rileykarson commented:
Sorry for the delay in getting to this! I'm aware and have it on a list of things to do, we're preparing our `3.0.0` release and I've been short on time. I'm hoping to get to it either this Friday or mid-late next week.

khanali21 commented:
@rileykarson  did you get a chance to review it ?

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 586be40d8da25e5cdec50785006248f3f4f1c833. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

rileykarson commented on products/appengine/terraform.yaml:
```diff
@@ -107,7 +107,31 @@ overrides: !ruby/object:Overrides::ResourceOverrides
+  ServiceSplitTraffic: !ruby/object:Overrides::Terraform::ResourceOverride
+    id_format: "{{service_id}}"
+    import_format: ["apps/{{project}}/services/{{service_id}}"]
+    mutex: "apps/{{project}}/services/{{service_id}}"
```
Most AppEngine operations need a mutex at the project level, `services/{{service_id}}` can be removed.

rileykarson commented on products/appengine/terraform.yaml:
```diff
@@ -107,7 +107,31 @@ overrides: !ruby/object:Overrides::ResourceOverrides
           bucket_name: "appengine-test-bucket"
-
+  ServiceSplitTraffic: !ruby/object:Overrides::Terraform::ResourceOverride
+    id_format: "{{service_id}}"
```
`id_format` and `import_format` are probably redundant with what's used by default. Did you find that you needed custom values?

rileykarson replied:
Resolving since this is really old.

rileykarson commented on products/appengine/terraform.yaml:
```diff
@@ -107,7 +107,31 @@ overrides: !ruby/object:Overrides::ResourceOverrides
+          bucket_name: "appengine-static-content"
+        test_env_vars:
+          org_id: :ORG_ID    
+        skip_test: true  
```
Do you remember why we had to skip adding a test? I think we can generate one.

khanali21 replied:
That's the problem, if I do not skip_test and generate it I get the error which I fixed using the work around. I have done quite a bit of debugging but could not reach to the bottom of this issue.

rileykarson replied:
Would you mind pushing a commit with the test enabled so I can take a look again? I'd really like to get it working.

rileykarson commented on products/appengine/api.yaml:
```diff
@@ -502,4 +502,74 @@ objects:
+      - !ruby/object:Api::Type::String
+        name: 'name'
+        description: |
+          Full path to the Service resource in the API. Example apps/myapp/services/default.
```
"The name of the service these settings apply to."

rileykarson commented on products/appengine/api.yaml:
```diff
@@ -502,4 +502,74 @@ objects:
+  - !ruby/object:Api::Resource
+    name: 'ServiceSplitTraffic'
+    description: |
+      A Service resource is a logical component of an application that can share state and communicate in a secure fashion with other services. 
```
"Traffic routing configuration for versions within a single service. Traffic splits define how traffic directed to the service is assigned to versions."

rileykarson commented on products/appengine/api.yaml:
```diff
@@ -502,4 +502,74 @@ objects:
+        description: |
+          Full path to the Service resource in the API. Example apps/myapp/services/default.
+          This field is used in responses only. Any value specified here in a request is ignored.
+      - !ruby/object:Api::Type::String
```
This seems redundant with `name`/`service`

rileykarson commented on products/appengine/api.yaml:
```diff
@@ -502,4 +502,74 @@ objects:
+          If set to true traffic will be migrated to this version.
+    properties:
+      - !ruby/object:Api::Type::String
+        name: 'name'
```
```suggestion
        name: 'service'
```

rileykarson commented on products/appengine/api.yaml:
```diff
@@ -502,4 +502,74 @@ objects:
+        properties:      
+          - !ruby/object:Api::Type::Enum
+            name: 'shardBy'
+            required: false
```
```suggestion
```

If not required, you can omit `required` entirely.

khanali21 commented:
I removed the skip_test and also removed the template for the test.
So now the test is being auto-generated.
I get this error:
```
root@84391ff50932:/mm# ./integration_tests.sh
==> Checking source code against gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google -v -run=TestAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample -timeout 240m -ldflags="-X=github.com/terraform-providers/terraform-provider-google/version.ProviderVersion=acc"
=== RUN   TestAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample
=== PAUSE TestAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample
=== CONT  TestAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample
--- FAIL: TestAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample (128.60s)
    testing.go:635: Step 1 error: Resource specified by ResourceName couldn't be found: google_app_engine_service_split_traffic.
FAIL
FAIL    github.com/terraform-providers/terraform-provider-google/google 128.629s
GNUmakefile:19: recipe for target 'testacc' failed
make: *** [testacc] Error 1
```

khanali21 commented:
> I removed the skip_test and also removed the template for the test.
> So now the test is being auto-generated.
> I get this error:
> 
> ```
> root@84391ff50932:/mm# ./integration_tests.sh
> ==> Checking source code against gofmt...
> ==> Checking that code complies with gofmt requirements...
> TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google -v -run=TestAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample -timeout 240m -ldflags="-X=github.com/terraform-providers/terraform-provider-google/version.ProviderVersion=acc"
> === RUN   TestAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample
> === PAUSE TestAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample
> === CONT  TestAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample
> --- FAIL: TestAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample (128.60s)
>     testing.go:635: Step 1 error: Resource specified by ResourceName couldn't be found: google_app_engine_service_split_traffic.
> FAIL
> FAIL    github.com/terraform-providers/terraform-provider-google/google 128.629s
> GNUmakefile:19: recipe for target 'testacc' failed
> make: *** [testacc] Error 1
> ```

I had debugged this and somehow could not understand the issue. The work around I found was to add the expected error as the regex.

rileykarson commented on products/appengine/api.yaml:
```diff
@@ -515,4 +515,70 @@ objects:
+    references: !ruby/object:Api::Resource::ReferenceLinks
+      api: 'https://cloud.google.com/appengine/docs/admin-api/reference/rest/v1/apps.services'
+    async: !ruby/object:Api::Async
+      operation: !ruby/object:Api::Async::Operation
```
```suggestion
      operation: !ruby/object:Api::OpAsync::Operation
```

`Async` was renamed `OpAsync`, do you mind updating that? Sorry for the delay in getting back!

khanali21 replied:
Thanks @rileykarson.

rileykarson commented on products/appengine/api.yaml:
```diff
@@ -515,4 +515,70 @@ objects:
+    update_mask: true
+    references: !ruby/object:Api::Resource::ReferenceLinks
+      api: 'https://cloud.google.com/appengine/docs/admin-api/reference/rest/v1/apps.services'
+    async: !ruby/object:Api::Async
```
I think all the occurrences of `::Async` here need to be `::OpAsync`, but I'm not sure. Can you amend them and make sure that it compiles locally?

khanali21 replied:
hmm... I was able to compile. But let me check after rebase.

khanali21 replied:
@rileykarson I went on long vacation, came back.
I have checked in the latest code with the rebase on the current master.
Everything fine but I am still getting the following error:
root@52d8e96b9325:/mm# ./integration_tests.sh
==> Checking source code against gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google -v -run=TestAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample -timeout 240m -ldflags="-X=github.com/terraform-providers/terraform-provider-google/version.ProviderVersion=acc"
=== RUN   TestAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample
=== PAUSE TestAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample
=== CONT  TestAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample
--- FAIL: TestAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample (136.23s)
    testing.go:635: Step 1 error: Resource specified by ResourceName couldn't be found: google_app_engine_service_split_traffic.
FAIL
FAIL    github.com/terraform-providers/terraform-provider-google/google 136.256s
GNUmakefile:19: recipe for target 'testacc' failed
make: *** [testacc] Error 1

rileykarson replied:
Replied in a comment below.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-2269-old..auto-pr-2269)
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-2269-old..auto-pr-2269)
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-2269-old..auto-pr-2269)

rileykarson commented on products/appengine/terraform.yaml:
```diff
@@ -109,7 +109,30 @@ overrides: !ruby/object:Overrides::ResourceOverrides
+    examples:
+      - !ruby/object:Provider::Terraform::Examples
+        name: "app_engine_service_split_traffic"
+        vars:
```
```suggestion
        primary_resource_id: 'myapp'
        vars:
```

This should resolve the issue you were seeing. It was caused when the test framework effectively ran `terraform import google_app_engine_service_split_traffic. {{id}}` instead of `terraform import google_app_engine_service_split_traffic.myapp {{id}}`

khanali21 replied:
Cool, thanks I have committed the suggestion.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-2269-old..auto-pr-2269) ( 5 files changed, 2 insertions(+), 672 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-2269-old..auto-pr-2269) ( 5 files changed, 2 insertions(+), 672 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-2269-old..auto-pr-2269) ( 1 file changed, 111 deletions(-))

rileykarson commented on products/appengine/terraform.yaml:
```diff
@@ -109,7 +109,31 @@ overrides: !ruby/object:Overrides::ResourceOverrides
           bucket_name: "appengine-test-bucket"
-
+  ServiceSplitTraffic: !ruby/object:Overrides::Terraform::ResourceOverride
+    id_format: "{{service_id}}"
```
Can you try removing the `id_format` and `import_format`? If they need to be there, can you comment with what the format is when they're not set?

We try to make `id` values look like `apps/{{project}}/services/{{service_id}}` now, so that should be the default behaviour. If the generated ones are wrong, you can use that form for both `id` and `import` I think.

khanali21 replied:
I tried. But it fails if I try to remove the the id_format and import_format , or change the id_format to apps/{{project}}/services/{{service_id}}. I am doing more testing to find out the reason for that.

rileykarson replied:
Can post the error you get when setting `id_format` to `['apps/{{project}}/services/{{service_id}}']`? Using these values is fine if the defaults aren't working.

khanali21 replied:
@rileykarson 
It is fixed. Don't get any error with id_format as "apps/{{project}}/services/{{service_id}}".
It was my bad.

rileykarson commented on templates/terraform/examples/app_engine_service_split_traffic.tf.erb:
```diff
@@ -0,0 +1,69 @@
+  }
+  depends_on = ["google_app_engine_standard_app_version.myapp_v2"]
+}
+resource "google_app_engine_service_split_traffic" "myapp2" {
```
I think we can remove this entry, having just the first is fine.

rileykarson commented on templates/terraform/examples/app_engine_service_split_traffic.tf.erb:
```diff
@@ -0,0 +1,69 @@
+      v2 = 0.25
+    }
+  }
+  depends_on = ["google_app_engine_standard_app_version.myapp_v2"]
```
This is redundant with the interpolation in `service_id` I think, and can be removed.

khanali21 replied:
This is needed in order to get the integration tests passed; as the test fail because it can not destroy the service which has zero traffic.

rileykarson replied:
I think the same connection between resources is established by https://github.com/GoogleCloudPlatform/magic-modules/pull/2269/files#diff-458f601ca7d7a40a89cb7b8bf8595b19R49, though?

khanali21 replied:
@rileykarson 
So I tested it , here is what is happening. We create v1 it takes 100% of the traffic, we create v2 it has no traffic, we use split traffic resource to create 75-25% split between the two versions.
Now when the destroy happens, it tries to destroy v1 which is having 75% traffic, and it fails as the v1 is part of split traffic, I don't think it automatically transfers the traffic to other v1.

What happens when I add the second split, it migrate all the traffic to v2, then destroy works fine.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-2269-old..auto-pr-2269) ( 5 files changed, 652 insertions(+), 2 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-2269-old..auto-pr-2269) ( 5 files changed, 652 insertions(+), 2 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-2269-old..auto-pr-2269) ( 1 file changed, 111 insertions(+))

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-2269-old..auto-pr-2269) ( 5 files changed, 676 insertions(+), 2 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-2269-old..auto-pr-2269) ( 5 files changed, 676 insertions(+), 2 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-2269-old..auto-pr-2269) ( 1 file changed, 111 insertions(+))

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-2269-old..auto-pr-2269) ( 5 files changed, 676 insertions(+), 2 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-2269-old..auto-pr-2269) ( 5 files changed, 676 insertions(+), 2 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-2269-old..auto-pr-2269) ( 1 file changed, 111 insertions(+))

rileykarson commented on templates/terraform/resource.erb:
```diff
@@ -131,9 +131,6 @@ func resource<%= resource_name -%><%= prop.name.camelize(:upper) -%>SetStyleDiff
 <% end -%>
 
 func resource<%= resource_name -%>Create(d *schema.ResourceData, meta interface{}) error {
-<%  if object.custom_code.custom_create -%>
```
I think I said to remove this, but it looks already committed. Never mind in that case- can you revert the changes to this file?

rileykarson replied:
These changes still need to be removed.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-2269-old..auto-pr-2269) ( 5 files changed, 619 insertions(+), 2 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-2269-old..auto-pr-2269) ( 5 files changed, 619 insertions(+), 2 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-2269-old..auto-pr-2269) ( 1 file changed, 91 insertions(+))

khanali21 commented:
@rileykarson Apologies for my complacency , I am starting getting the following error after applying the changes
==> Checking that code complies with gofmt requirements...
go install
# github.com/terraform-providers/terraform-provider-google/google
google/error_retry_predicates.go:62:23: urlerr.Unwrap undefined (type *url.Error has no field or method Unwrap)
google/error_retry_predicates.go:76:24: urlerr.Unwrap undefined (type *url.Error has no field or method Unwrap)
GNUmakefile:11: recipe for target 'build' failed
make: *** [build] Error 2

I am trying to dig it.

khanali21 commented:
@rileykarson I am struggling with the following error, apparently this is due to a change 
Do I need to update something?


commit c94b6c7206d4bba55909c0ed84b83560a7f4bbc3
Author: The Magician <magic-modules@google.com>
Date:   Thu Feb 13 12:29:36 2020 -0800

    Add more generic error predicates (#3116) (#5675)

    * add connection reset by peer predicate

    * lint

    * separate error preds, fix temp pred

    * Fix error retries, comments

    * remove removed func

    Signed-off-by: Modular Magician <magic-modules@google.com>




 
==> Checking that code complies with gofmt requirements...
go install
# github.com/terraform-providers/terraform-provider-google/google
google/error_retry_predicates.go:62:23: urlerr.Unwrap undefined (type *url.Error has no field or method Unwrap)
google/error_retry_predicates.go:76:24: urlerr.Unwrap undefined (type *url.Error has no field or method Unwrap)
GNUmakefile:11: recipe for target 'build' failed
make: *** [build] Error 2

khanali21 commented:
@rileykarson Still struggling with it, I have tried even with the old code that was passing the compilation. I have also tried reverting the changes in module terraform provider.
Any help is appreciated. Thanks in advance.

rileykarson commented on products/appengine/api.yaml:
```diff
@@ -515,4 +515,69 @@ objects:
+    properties:
+      - !ruby/object:Api::Type::String
+        name: 'service'
+        url_param_only: true
```
```suggestion
        api_name: 'id'
```

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-2269-old..auto-pr-2269) ( 5 files changed, 647 insertions(+), 2 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-2269-old..auto-pr-2269) ( 5 files changed, 647 insertions(+), 2 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-2269-old..auto-pr-2269) ( 1 file changed, 101 insertions(+))

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-2269-old..auto-pr-2269) ( 20 files changed, 697 insertions(+), 94 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-2269-old..auto-pr-2269) ( 22 files changed, 705 insertions(+), 113 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-2269-old..auto-pr-2269) ( 1 file changed, 101 insertions(+))

khanali21 commented:
@rileykarson DONE :) Had a bad time with this one.... even the changes from resource.erb were not going away git acting up with resolving merge conflict.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-2269-old..auto-pr-2269) ( 5 files changed, 647 insertions(+), 2 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-2269-old..auto-pr-2269) ( 5 files changed, 647 insertions(+), 2 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-2269-old..auto-pr-2269) ( 1 file changed, 101 insertions(+))
